### DISCUSSION TITLE
scikit-build f2py compile error with full path of signature file on Windows: Perhaps anyone here can help out?

### DISCUSSION THREAD

**Discussion by jcfr** [scikit-build/scikit-build#822] [Q&A]
_Originally posted by @jmrohwer in https://groups.google.com/g/scikit-build/c/L7SmriBFeFk/m/0mIeunr_BAAJ_

---

Hi all,

I am trying to port the build system for a python package ([GitHub - PySCeS/pysces: The official PySCeS project source code repository.](http://github.com/PySCeS/pysces)) from `numpy.distutils` to use `scikit-build` because of the deprecation of `distutils` (the package contains 2 Fortran extension modules wrapped with `f2py`). To get to know the system I’ve started with a minimal example that I've put together from reading the numpy/f2py and scikit-build documentation. I have been getting compile failures on Windows (on Linux and macOS everything works, I can build and install the module).

I have posted this on SO and on the Fortran discourse group but have not had any replies despite quite a number of reads. Perhaps anyone here can help out? Links to the other posts:

* [python - f2py compile error with signature file on Windows - Stack Overflow](https://stackoverflow.com/questions/73464905/f2py-compile-error-with-signature-file-on-windows)
* [Fortran language discourse group](https://fortran-lang.discourse.group/t/f2py-compile-error-with-full-path-of-signature-file-on-windows/4211)

Minimal example (I attach the zipped full source tree of the package for reference if needed. See [simple.zip](https://github.com/scikit-build/scikit-build/files/10232166/simple.zip)):

**hi.f**

```
      SUBROUTINE HELLO()
      PRINT*,"Hello from fortran"
      END
```

**hi.pyf**

```
!    -*- f90 -*-
! Note: the context of this file is case sensitive.

python module hi ! in 
    interface  ! in :hi
        subroutine hello ! in :hi:hi.f
        end subroutine hello
    end interface 
end python module hi

! This file was auto-generated with f2py (version:1.23.2).
! See:
! https://web.archive.org/web/20140822061353/http://cens.ioc.ee/projects/f2py2e
```

During compilation with `python setup.py build` the build fails. In tracking it down I saw that ninja/cmake expands the full path of the source files and this is where it fails. To ensure that it is not a problem with the `skbuild` build system I verified that the failure can be reproduced on the command line using `f2py` on its own as follows - copy the Fortran files above to the same folder and then run in a command prompt:

```
(skbuild) Z:\Documents\Python\skbuild_test\simple\hello\shout>f2py.exe -m hi -c hi.pyf hi.f
```

(works and generates the `.pyd` and `.dll` files)

However, specifying the full path, fails:

```
(skbuild) Z:\Documents\Python\skbuild_test\simple\hello\shout>f2py.exe -m hi -c Z:/Documents/Python/skbuild_test/simple/hello/shout/hi.pyf Z:/Documents/Python/skbuild_test/simple/hello/shout/hi.f
running build
running config_cc
INFO: unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
INFO: unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
INFO: build_src
INFO: building extension "hi" sources
creating C:
creating C:Users
creating C:Users\jr
creating C:Users\jr\AppData
creating C:Users\jr\AppData\Local
creating C:Users\jr\AppData\Local\Temp
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10\Documents
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10\Documents\Python
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10\Documents\Python\skbuild_test
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10\Documents\Python\skbuild_test\simple
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10\Documents\Python\skbuild_test\simple\hello
creating C:Users\jr\AppData\Local\Temp\tmpdbudw6jh\src.win-amd64-3.10\Documents\Python\skbuild_test\simple\hello\shout
INFO: f2py options: []
INFO: f2py: Z:/Documents/Python/skbuild_test/simple/hello/shout/hi.pyf
error: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:C:\\Users\\jr\\AppData\\Local\\Temp\\tmpdbudw6jh\\src.win-amd64-3.10\\Documents\\Python\\skbuild_test\\simple\\hello\\shout'
```

The problem seems to be with the `creating C:Users` part leading to the malformed filename in the last line.

Specifying only the filename for `hi.pyf` but the full path for `hi.f` **does work** strangely enough, but as soon as the full path for `hi.pyf` is specified it fails. As mentioned, on Linux and macOS all works fine.

---

**NOTE:**
I could of course skip the `.pyf` file and then it works. However, the Fortran code I’m wrapping is F77 and needs to be redistributed in unmodified form due to licensing conditions, so I do need a custom `.pyf` file to integrate it into my Python module.

---

**System**

* Windows 10 64-bit
* MSVC 2022 build tools
* mingw 11.2.0.07112021
* python 3.10.6
* numpy 1.23.2

I realize this may not be directly related to skbuild, since the error can be reproduced on the command line using only f2py, but perhaps someone here has seen anything similar? I honestly have no idea what is going on here.

Thanks,
Johann

**Comment by jmrohwer** [comment]
@jcfr Thanks for moving this discussion. To follow up, I got this to work by manually specifying `--build-dir` in the f2py command, then it works on all platforms:
https://github.com/PySCeS/pysces/blob/416bd70dcbac779bec3c99b3ab1ff78b4fc34c91/pysces/nleq2/CMakeLists.txt#L21-L28

See also https://github.com/numpy/numpy/issues/22211

**Comment by jmrohwer** [comment]
Incidentally as part of the transition away from distutils, numpy's `f2py-c` is being ported to MESON:
https://github.com/numpy/numpy/pull/22225

However, this does not work (yet) with custom `*.pyf` files:
https://github.com/numpy/numpy/pull/22225#issuecomment-1250874383

**Comment by henryiii** [comment]
I think the latest versions of f2py-cmake work a lot better.
